Molecule testing with optional audit using CISCAT Pro

bbaassssiiee opened the pull request:
### validating the compliance with auditing

To implement this I started with a test framework called Molecule v1, also to improve code quality and testability in general.

Fixing some trivial issues along the way, and adding just a few workarounds I reached my higher goal of validating of this Ansible role using the tool that CIS provides to its members: **ciscat-full-bundle-2017-11-06-v3.0.44.zip**

**to create tests/report.html run:**
```
make test
```


CISCAT Pro is [distributed independently ](https://workbench.cisecurity.org/files/1816/download) It should be downloaded to /usr/local/share/CIS.

### For better quality we should add tests that can be run everywhere.
- [x] Create virtualenv for Ansible and use pip to install python modules.
- [ ] Check Yaml style (using yamllint)
- [x] Check Ansible best practices (using molecule)
- [x] Run the role on an official Centos/7 box
- [x] Check idempotence (using molecule)
- [x] Run an independent audit for compliance
- [x] Run all the tests with one command (make test)
- [x] Run all the tests under docker
- [x] Minimize Travis magic

bbaassssiiee committed:
adding Molecule testing

bbaassssiiee committed:
pinning to a stable and simple version of Molecule.

bbaassssiiee committed:
removing trailing newlines

bbaassssiiee committed:
adding tag skip_ansible_lint for specific commands etc.

bbaassssiiee committed:
adding a group for later testing

bbaassssiiee committed:
molecule verify will use CISCAT Pro (CIS members only)

bbaassssiiee committed:
use command instead of shell when you can

bbaassssiiee committed:
Adding temp files to gitignore

bbaassssiiee committed:
How to use CISCAT Pro if you are a CIS Member

bbaassssiiee committed:
adding one "changed_when: no" for idempotence

bbaassssiiee committed:
exclude CISCAT audit reports from Git

shepdelacreme reviewed (changes requested):
I don't see a reason to remove the Dockerfile. Also it breaks the testing we have in place using Travis CI. I'm generally fine moving to use this Molecule framework for testing but it needs to continue to work with Travis or another automated CI system since I don't want to rely on local testing only. 

It looks like currently the testing is entirely centered around running against a vagrant box...which I generally prefer but unfortunately I haven't found a CI vendor that you can easily use vagrant boxes with...instead of docker images.

Also the testing should be able to complete successfully without the use of CISCAT Pro. Membership is not cheap.

bbaassssiiee committed:
fix for #50

bbaassssiiee committed:
fix for #55

bbaassssiiee committed:
changes suggested by @shepdelacreme

bbaassssiiee committed:
:dizzy_face: Dockerfile was removed unintentionally.

bbaassssiiee committed:
The testing will complete successfully without the use of CISCAT Pro.

CIS Pro membership is not cheap. The playbook tests/test_default.yml will skip the benchmark unless you are member.
Only as a (corporate) member you are allowed to use /usr/local/share/CIS/ciscat-full-bundle-2017-11-06-v3.0.44.zip
In that case the tests/report.html and tests/report.txt will help you resolve issues.

bbaassssiiee committed:
make help

bbaassssiiee committed:
fixes workaround

bbaassssiiee committed:
fixes another workaround

bbaassssiiee committed:
enabling CISCAT when available

bbaassssiiee committed:
DRY

bbaassssiiee committed:
running molecule under docker

bbaassssiiee committed:
molecule runs under docker now.

bbaassssiiee committed:
removed action not needed

bbaassssiiee committed:
copy molecule.yml from vagrant.yml or docker.yml (default).

bbaassssiiee committed:
fixed group

bbaassssiiee committed:
run the audit on level 2

bbaassssiiee committed:
compare docker vs virtualbox implementation

shepdelacreme reviewed (changes requested):
There is one specific requested change in this review. Also if you are mostly done with changes to this PR I'm going to pull it and test out the Molecule stuff locally.

bbaassssiiee committed:
This changed_when: no line is not needed anymore.

bbaassssiiee committed:
ansible.cfg fixes permission problem in idempotency run

bbaassssiiee committed:
use official Centos image

bbaassssiiee committed:
use official Centos image

bbaassssiiee committed:
drastically reduced .travis.yml to avoid too much dependency on Travis

shepdelacreme committed:
Merge branch 'devel' into devel

shepdelacreme committed:
Merge branch 'devel' into devel
